### PR TITLE
SCUMM: GUI: Implement MI1 SegaCD menu

### PR DESCRIPTION
**Pull request by AndywinXp**
Thanks to @einstein95, I've been able to fully implement the menu for the Sega CD version of The Secret Of Monkey Island.

F5 will open the GMM, while the SPACE key, acting as our pause button, opens the menu just like on the original.
The SegaCD apparently has no banners (I haven't been able to trigger those...), so those are disabled here as well.

The passcode saving/loading system is fully functional, but please give it some further testing if you can 🙂 

Here are the usual screenshots:

image
image
image
image
image

### CODE REVIEW

File: engines/scumm/gfx_gui.cpp [hunk 1]
```diff
@@ -1964,7 +1991,163 @@ void ScummEngine::showMainMenu() {
...
+	case GUI_CTRL_NUMPAD_3:
+	case GUI_CTRL_NUMPAD_4:
+	case GUI_CTRL_NUMPAD_5:
+	case GUI_CTRL_NUMPAD_6:
+	case GUI_CTRL_NUMPAD_7:
+	case GUI_CTRL_NUMPAD_8:
+	case GUI_CTRL_NUMPAD_9:
+	case GUI_CTRL_NUMPAD_BACK:
+	{
+		int inputNum = op == GUI_CTRL_NUMPAD_0 ? 0 : op;
```
Comment (sev-) [inline engines/scumm/gfx_gui.cpp:2108]: please put parentheses for better readability: 
```C
int inputNum = (op == GUI_CTRL_NUMPAD_0) ? 0 : op;
```

File: engines/scumm/gfx_gui.cpp [hunk 2]
```diff
@@ -1964,7 +1991,163 @@ void ScummEngine::showMainMenu() {
...
+		uint curIdx;
+		for (curIdx = 0; curIdx < sizeof(_mainMenuSegaCDPasscode); curIdx++) {
```
Comment (sev-) [inline engines/scumm/gfx_gui.cpp:2110]: could it be 
```C
curIdx = strnlen(_mainMenuSegaCDPasscode, sizeof(_mainMenuSegaCDPasscode));
```

or even `strlen()` as you have 8 as max length anyway.

File: engines/scumm/gfx_gui.cpp [hunk 3]
```diff
@@ -1964,7 +1991,163 @@ void ScummEngine::showMainMenu() {
...
+			if (curIdx > 0) {
+				_mainMenuSegaCDPasscode[curIdx - 1] = '\0';
+			}
+
+		} else {
+			_mainMenuSegaCDPasscode[curIdx] = '0' + inputNum;
+
+			if (curIdx >= 3) { // Last digit
+				updateMainMenuControls();
+				ScummEngine::drawDirtyScreenParts();
```
Comment (sev-) [inline engines/scumm/gfx_gui.cpp:2125]: You're in the `ScummEngine` class anyway. The namespace is redundant here if I understand it correctly.

File: engines/scumm/gfx_gui.cpp [hunk 4]
```diff
@@ -2559,6 +2748,217 @@ void ScummEngine::setUpMainMenuControls() {
...
+								getBannerColor(20),
+								getBannerColor(19),
+								getBannerColor(10),
+								getBannerColor(12),
+								152,
+								yConstant + 18,
+								235,
+								yConstant + 34,
+								_uncheckedBox, 1, 1);
+	}
```
Comment (sev-) [inline engines/scumm/gfx_gui.cpp:2872]: Maybe add `return;` here?

File: engines/scumm/gfx_gui.cpp [hunk 5]
```diff
@@ -2559,6 +2748,217 @@ void ScummEngine::setUpMainMenuControls() {
...
+								getBannerColor(7),
+								206,
+								yConstant + 21,
+								isJap ? 291 : 257,
+								yConstant + 43,
+								getGUIString(gsCancel), 1, 1);
+	}
+
+	if (_menuPage == GUI_PAGE_LOAD) {
+		Common::String numbers[] = {"0", "1", "2", "3", "4", "5", "6", "7", "8", "9"};
```
Comment (sev-) [inline engines/scumm/gfx_gui.cpp:2909]: `const char *numbers[2] = ....`

Same effect but skips the instantiation of 10 objects

File: engines/scumm/gfx_gui.cpp [hunk 6]
```diff
@@ -2928,6 +3328,11 @@ void ScummEngine_v6::setUpMainMenuControls() {
 }
 
 void ScummEngine::drawMainMenuControls() {
+	if (_game.platform == Common::kPlatformSegaCD) {
```
Comment (sev-) [inline engines/scumm/gfx_gui.cpp:3331]: For a moment, I was thinking about subclassing `ScummEngine` to `ScummEngineSEGA` and overriding these methods there.

File: engines/scumm/gfx_gui.cpp [hunk 7]
```diff
@@ -3130,7 +3613,43 @@ void ScummEngine::updateMainMenuControls() {
 	_system->updateScreen();
 }
 
+void ScummEngine::updateMainMenuControlsSegaCD() {
+	char msg[256];
+	int yConstant = _virtscr[kMainVirtScreen].topline + (_virtscr[kMainVirtScreen].h / 2);
+	bool isJap = _language == Common::JA_JPN;
```
Comment (sev-) [inline engines/scumm/gfx_gui.cpp:3619]: Please put parentheses for better readability

File: engines/scumm/gfx_gui.cpp [hunk 8]
```diff
@@ -3130,7 +3613,43 @@ void ScummEngine::updateMainMenuControls() {
...
+
+	if (_menuPage == GUI_PAGE_MAIN) {
+		strncpy(_mainMenuTextSpeedSlider, "\x3a\x3a\x3a\x3a\x3a\x3a\x3a\x3a\x3a\x3a", sizeof(_mainMenuTextSpeedSlider));
```
Comment (sev-) [inline engines/scumm/gfx_gui.cpp:3622]: Please describe the string in a comment. Something like `// Horizontal line`

Same for `\x3b` below

File: engines/scumm/input.cpp [hunk 9]
```diff
@@ -897,20 +898,26 @@ void ScummEngine::processKeyboard(Common::KeyState lastKeyHit) {
...
-			CursorMan.showMouse(_cursor.state > 0);
-			// "Game Paused.  Press SPACE to Continue."
-			if (_game.version > 4)
-				showBannerAndPause(0, -1, getGUIString(gsPause));
-			else
-				showOldStyleBannerAndPause(getGUIString(gsPause), 12, -1);
-
-			_cursor.state = oldCursorState;
-			return;
+		if (((VAR_PAUSE_KEY != 0xFF && (lastKeyHit.ascii == VAR(VAR_PAUSE_KEY))) ||
```
Comment (sev-) [inline engines/scumm/input.cpp:901]: I wonder why did you add two extra sets of parentheses here. Looks unnecessary to me.

File: engines/scumm/scumm.h [hunk 10]
```diff
@@ -444,7 +462,13 @@ enum GUIString {
...
 	gsCGAMode = 53,
 	gsHerculesMode = 54,
-	gsTandyMode = 55
+	gsTandyMode = 55,
+	gsCurrentPasscode = 56,
+	gsEnterPasscode = 57,
+	gsConfirmPasscode = 58,
+	gsInvalidPasscode = 59,
+	gsSlowFast = 60,
+	gsRestartGame = 61
```
Comment (sev-) [inline engines/scumm/scumm.h:471]: Add a comma here, we live in c++11 (yay!)

**Review by sev- — COMMENTED** [review]
It is great to see these still coming!

I added several notes.

File: engines/scumm/gfx_gui.cpp (hunk 3 above)
Comment (AndywinXp) [reply engines/scumm/gfx_gui.cpp:2125]: You are correct, thanks; I was doing this out of habit of handling multiple SCUMM versions, but this is v5 and doesn't need the namespace explicitly defined in the call.

File: engines/scumm/gfx_gui.cpp (hunk 4 above)
Comment (AndywinXp) [reply engines/scumm/gfx_gui.cpp:2872]: Whoops! Yes, given the purpose of the code I'd say that changing the other `if`s to `else if` would improve readability and underline the fact that the menu pages are mutually exclusive to each other.

File: engines/scumm/gfx_gui.cpp (hunk 5 above)
Comment (AndywinXp) [reply engines/scumm/gfx_gui.cpp:2909]: Thanks!

File: engines/scumm/input.cpp (hunk 9 above)
Comment (AndywinXp) [reply engines/scumm/input.cpp:901]: A result of me coding at night :-P thanks!

File: engines/scumm/gfx_gui.cpp (hunk 6 above)
Comment (AndywinXp) [reply engines/scumm/gfx_gui.cpp:3331]: Uhm, interesting. I'll see what would the effort of that be in terms of moving stuff around to another subclass; it's possible that that would have to wait for a different PR or a commit external to this one.

File: engines/scumm/gfx_gui.cpp (hunk 2 above)
Comment (AndywinXp) [reply engines/scumm/gfx_gui.cpp:2110]: Definitely! Thanks!

**Comment by AndywinXp** [comment]
Thanks sev! I'm thinking that the subclass one might involve moving an amount of stuff around between `ScummEngine_v5`, `ScummEngine` and the hypothetical `ScummEngineSEGA`, something which might better be suited for another PR or external commit.

**Review by sev- — APPROVED** [review]

**Comment by sev-** [comment]
I have no further notes. Merging.
